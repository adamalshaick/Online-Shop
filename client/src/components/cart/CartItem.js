import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { removeItemFromCart } from "../../actions/cartActions";
import { getProfileById } from "../../actions/profileActions";
import { Link } from "react-router-dom";

export class CartItem extends Component {
  componentDidMount() {
    this.props.getProfileById(this.props.item.user);
  }

  onDeleteClick(id) {
    this.props.removeItemFromCart(id);
  }

  render() {
    const { item, loading } = this.props;
    const { profileById } = this.props.profile;
    return (
      <div style={{ width: "100%" }}>
        <img
          style={{ height: "50px", width: "50px" }}
          // src={`../../uploads/post_image/${item.itemImage}`}
          src="../../uploads/post_image/placeholder.png"
          alt=""
        />
        {item.title}

        <button
          onClick={this.onDeleteClick.bind(this, item._id)}
          type="button"
          className="btn btn-danger ml-3 float-right"
        >
          Remove
        </button>

        {profileById === null || loading ? null : (
          <Link
            to={`/profile/${profileById.handle}`}
            className="btn btn-light float-right ml-3 d-none d-sm-block"
          >
            Seller's Profile
          </Link>
        )}

        <span className="mt-2 float-right">{item.price} $</span>
        <hr />
      </div>
    );
  }
}

CartItem.propTypes = {
  item: PropTypes.object.isRequired,
  removeItemFromCart: PropTypes.func.isRequired,
  profile: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  auth: state.auth,
  profile: state.profile
});

export default connect(
  mapStateToProps,
  { removeItemFromCart, getProfileById }
)(CartItem);
