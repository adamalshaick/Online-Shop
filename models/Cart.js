const mongoose = require("mongoose");
const Schema = mongoose.Schema;

//Create Schema
const CartSchema = new Schema({
  id: {
    type: String
  },

  cart: [
    {
      type: Schema.ObjectId,
      ref: "item"
    }
  ]
});

module.exports = Cart = mongoose.model("cart", CartSchema);
