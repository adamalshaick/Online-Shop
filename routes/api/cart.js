const express = require("express");
const router = express.Router();
const mongoose = require("mongoose");
const passport = require("passport");

// Item model
const Item = require("../../models/Item");

// Card model
const Cart = require("../../models/Cart");

// @route   POST api/cart
// @desc    Add item to cart
// @access  Private
router.post(
  "/",
  passport.authenticate("jwt", { session: false }),
  (req, res) => {
    // const { errors, isValid } = validateCartInput(req.body);

    // Check Validation
    // if (!isValid) {
    //   // If any errors, send 400 with errors object
    //   return res.status(400).json(errors);
    // }

    Cart.findOne({ user: req.user.id })
      .then(cart => {
        if (cart) {
          cart.items.unshift({ item: req.body });
          cart.value = +card.value + +req.body.price;
          cart.save().then(card => res.json(card));
        } else {
          const cartFields = {
            user: req.user.id,
            items: [{ item: req.body }],
            value: req.body.price
          };

          new Cart(cartFields).save().then(cart => res.json(cart));
        }
      })

      .catch(err => {
        res.json(err);
      });
  }
);

// @route   GET api/cart
// @desc    Get items from cart
// @access  Private
router.get(
  "/",
  passport.authenticate("jwt", { session: false }),
  (req, res) => {
    Cart.findOne({ user: req.user.id })
      .then(cart => {
        res.json(cart);
      })
      .catch(err => {
        res.json(err);
      });
  }
);

module.exports = router;
