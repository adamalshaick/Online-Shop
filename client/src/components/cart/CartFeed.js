import React, { Component } from "react";
import PropTypes from "prop-types";
import CartItem from "./CartItem";
import { Header } from "../common/styles/Header";
import { connect } from "react-redux";
import { buyItems } from "../../actions/cartActions";

export class CartFeed extends Component {
  render() {
    const { cartItems } = this.props;
    let cartValue = 0;
    cartItems.forEach(item => {
      cartValue = +item.price + cartValue;
    });
    return (
      <div className="entry">
        <Header>
          <span style={{ fontSize: "18px", float: "left" }}>
            <i className="mr-2 fas fa-shopping-cart" />
            <strong>Your cart</strong>
          </span>
          <span style={{ fontSize: "18px", float: "right" }}>
            <strong>Amount: {cartValue} $</strong>
          </span>
        </Header>

        {cartItems.map(item => (
          <CartItem key={item._id} item={item} />
        ))}
        <div className="float-right mb-5">
          <div>
            <small>(Not working yet, about to be implemented)</small>
          </div>
          <button
            type="button"
            onClick={this.onClick}
            className="btn btn-success float-right mb-5"
          >
            Buy Items <i className="fas fa-arrow-circle-right ml-2" />
          </button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  auth: state.auth
});

CartFeed.propTypes = {
  cartItems: PropTypes.array.isRequired,
  auth: PropTypes.object.isRequired,
  buyItems: PropTypes.func.isRequired
};

export default connect(
  mapStateToProps,
  { buyItems }
)(CartFeed);
