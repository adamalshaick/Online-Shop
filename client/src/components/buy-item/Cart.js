import React, { Component } from "react";
import { connect } from "react-redux";
import { getItemsFromCart } from "../../actions/itemActions";
import CartItem from "./CartItem";
import CartFeed from "./CartFeed";

class Cart extends Component {
  componentDidMount() {
    this.props.getItemsFromCart();
  }

  render() {
    const { cart, loading } = this.props.cart;

    let itemContent;
    if (cart === null || loading) {
      itemContent = <p>Loading ...</p>;
    } else {
      itemContent = <CartFeed cartItems={cart} />;
    }

    return <div>{itemContent}</div>;
  }
}

const mapStateToProps = state => ({
  cart: state.cart
});

export default connect(
  mapStateToProps,
  { getItemsFromCart }
)(Cart);
