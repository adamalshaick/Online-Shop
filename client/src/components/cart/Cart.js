import React from "react";
import fetchCurrentUser from "../common/hoc/fetchCurrentUser";
import CartFeed from "./CartFeed";
import PropTypes from "prop-types";

export const Cart = ({ user }) => {
  return (
    <div className="container">
      <CartFeed cartItems={user.currentUser.cart} />
    </div>
  );
};

Cart.propTypes = {
  user: PropTypes.object.isRequired
};

export default fetchCurrentUser(Cart);
