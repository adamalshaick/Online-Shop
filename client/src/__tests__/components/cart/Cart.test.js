import React from "react";
import { Cart } from "../../../components/cart/Cart";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

const mockCartfn = jest.fn();
const Props = {
  cart: {},
  loading: false
};

const wrapper = shallow(<Cart {...Props} getItemsFromCart={mockCartfn} />);
describe("render component", () => {
  it("renders component", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
