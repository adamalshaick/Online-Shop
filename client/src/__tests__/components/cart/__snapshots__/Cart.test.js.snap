// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render component renders component 1`] = `
<Fragment>
  <Connect(Navbar) />
  <div
    className="container"
  >
    <Connect(CartFeed) />
  </div>
</Fragment>
`;
